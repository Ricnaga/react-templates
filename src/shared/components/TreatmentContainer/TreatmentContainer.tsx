import { ReactElement } from 'react';

import { Grid, Skeleton, Stack, Text } from '@chakra-ui/react';

import { RequestStatus } from '@application/api/axios/useAxios';

interface TreatmentContainerProps<T> {
  data: T | undefined;
  isEmpty?: boolean;
  status?: keyof typeof RequestStatus;
  loadingCard?: ReactElement;
  emptyCard?: ReactElement;
  children: (data: T) => ReactElement;
}

export function TreatmentContainer<T>({
  data,
  status,
  isEmpty = false,
  children,
  emptyCard,
  loadingCard,
}: TreatmentContainerProps<T>) {
  switch (status) {
    case RequestStatus.FAIL:
      return (
        <Text fontSize="xl" textAlign="center">
          Erro! Não foi possível receber dados do backend
        </Text>
      );

    case RequestStatus.PENDING: {
      return (
        loadingCard || (
          <Stack>
            {Array.from({ length: 3 }).map(() => (
              <Skeleton key={Math.random()} height="1.5rem" />
            ))}
          </Stack>
        )
      );
    }

    default: {
      if (isEmpty) return emptyCard || <Grid>IS EMPTY</Grid>;

      return <Grid>{data && children(data)}</Grid>;
    }
  }
}
