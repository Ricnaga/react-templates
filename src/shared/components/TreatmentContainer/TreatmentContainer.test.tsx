import { Grid } from '@chakra-ui/react';

import { render } from '@application/test/testing-library';

import { TreatmentContainer } from './TreatmentContainer';

describe('Component: TreatmentContainer', () => {
  it('should render correctly', () => {
    const { container } = render(
      <TreatmentContainer data={{ id: 'mock' }}>
        {({ id }) => <Grid>{id}</Grid>}
      </TreatmentContainer>,
    );

    expect(container).toMatchSnapshot();
  });
});
