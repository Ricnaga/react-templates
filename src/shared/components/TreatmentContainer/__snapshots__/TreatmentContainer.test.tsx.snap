// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: TreatmentContainer should render correctly 1`] = `
<div>
  <div
    class="css-lgj0h8"
  >
    <div
      class="css-lgj0h8"
    >
      mock
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</div>
`;
